File: utils/kupola.go
package kupola

import (
	"fmt"
	"html/template"
)

func Button(label string, variant string, size string) template.HTML {
	if variant == "" {
		variant = "brand"
	}
	if size == "" {
		size = "md"
	}
	return template.HTML(fmt.Sprintf(
		`<button class="ds-btn ds-btn--%s ds-btn--%s">%s</button>`,
		variant, size, label,
	))
}

func Input(label string, name string, placeholder string, validate string) template.HTML {
	validateAttr := ""
	if validate != "" {
		validateAttr = fmt.Sprintf(` data-validate="%s"`, validate)
	}
	return template.HTML(fmt.Sprintf(
		`<div class="ds-form-item">
			<label class="ds-form-item__label">%s</label>
			<input type="text" class="ds-input" name="%s" placeholder="%s"%s>
		</div>`,
		label, name, placeholder, validateAttr,
	))
}

func Card(title string, body string, footer string) template.HTML {
	footerHTML := ""
	if footer != "" {
		footerHTML = fmt.Sprintf(`<div class="ds-card__footer">%s</div>`, footer)
	}
	return template.HTML(fmt.Sprintf(
		`<div class="ds-card">
			<div class="ds-card__header">
				<h3 class="ds-card__title">%s</h3>
			</div>
			<div class="ds-card__body">%s</div>
			%s
		</div>`,
		title, body, footerHTML,
	))
}

func StatCard(title string, value string, trend string, trendType string) template.HTML {
	trendHTML := ""
	if trend != "" {
		trendClass := "ds-stat-card__trend--up"
		if trendType == "down" {
			trendClass = "ds-stat-card__trend--down"
		}
		trendHTML = fmt.Sprintf(`<span class="ds-stat-card__trend %s">%s</span>`, trendClass, trend)
	}
	return template.HTML(fmt.Sprintf(
		`<div class="ds-stat-card">
			<div class="ds-stat-card__title">%s</div>
			<div class="ds-stat-card__value">%s</div>
			%s
		</div>`,
		title, value, trendHTML,
	))
}

func Badge(text string, variant string) template.HTML {
	if variant == "" {
		variant = "default"
	}
	return template.HTML(fmt.Sprintf(
		`<span class="ds-badge ds-badge--%s">%s</span>`,
		variant, text,
	))
}

func Avatar(initials string, size string, variant string) template.HTML {
	if size == "" {
		size = "md"
	}
	if variant == "" {
		variant = "default"
	}
	return template.HTML(fmt.Sprintf(
		`<div class="ds-avatar ds-avatar--%s ds-avatar--%s">%s</div>`,
		size, variant, initials,
	))
}

func ThemeScript(theme string, brand string) template.HTML {
	if theme == "" {
		theme = "dark"
	}
	if brand == "" {
		brand = "zengqing"
	}
	return template.HTML(fmt.Sprintf(
		`<script>
			if (typeof setTheme === 'function') setTheme('%s');
			if (typeof setBrand === 'function') setBrand('%s');
		</script>`,
		theme, brand,
	))
}

func DataBind(key string, bindType string) template.HTML {
	return template.HTML(fmt.Sprintf(`data-bind="%s:%s"`, key, bindType))
}

func DataComponent(name string) template.HTML {
	return template.HTML(fmt.Sprintf(`data-component="%s"`, name))
}

func DataValidate(rules string) template.HTML {
	return template.HTML(fmt.Sprintf(`data-validate="%s"`, rules))
}

File: utils/kupola.rb
module Kupola
  def self.button(label, variant = 'brand', size = 'md')
    "<button class=\"ds-btn ds-btn--#{variant} ds-btn--#{size}\">#{label}</button>"
  end

  def self.input(label, name, placeholder, validate = '')
    validate_attr = validate.empty? ? '' : " data-validate=\"#{validate}\""
    <<~HTML
      <div class="ds-form-item">
        <label class="ds-form-item__label">#{label}</label>
        <input type="text" class="ds-input" name="#{name}" placeholder="#{placeholder}"#{validate_attr}>
      </div>
    HTML
  end

  def self.card(title, body, footer = '')
    footer_html = footer.empty? ? '' : "<div class=\"ds-card__footer\">#{footer}</div>"
    <<~HTML
      <div class="ds-card">
        <div class="ds-card__header">
          <h3 class="ds-card__title">#{title}</h3>
        </div>
        <div class="ds-card__body">#{body}</div>
        #{footer_html}
      </div>
    HTML
  end

  def self.stat_card(title, value, trend = '', trend_type = 'up')
    trend_class = trend_type == 'down' ? 'ds-stat-card__trend--down' : 'ds-stat-card__trend--up'
    trend_html = trend.empty? ? '' : "<span class=\"ds-stat-card__trend #{trend_class}\">#{trend}</span>"
    <<~HTML
      <div class="ds-stat-card">
        <div class="ds-stat-card__title">#{title}</div>
        <div class="ds-stat-card__value">#{value}</div>
        #{trend_html}
      </div>
    HTML
  end

  def self.badge(text, variant = 'default')
    "<span class=\"ds-badge ds-badge--#{variant}\">#{text}</span>"
  end

  def self.avatar(initials, size = 'md', variant = 'default')
    "<div class=\"ds-avatar ds-avatar--#{size} ds-avatar--#{variant}\">#{initials}</div>"
  end

  def self.theme_script(theme = 'dark', brand = 'zengqing')
    <<~HTML
      <script>
        if (typeof setTheme === 'function') setTheme('#{theme}');
        if (typeof setBrand === 'function') setBrand('#{brand}');
      </script>
    HTML
  end

  def self.data_bind(key, bind_type)
    "data-bind=\"#{key}:#{bind_type}\""
  end

  def self.data_component(name)
    "data-component=\"#{name}\""
  end

  def self.data_validate(rules)
    "data-validate=\"#{rules}\""
  end
end